File: couchbase-entity/src/main/java/com/kaufland/generation/TypeConversionMethodsGeneration.java
package com.kaufland.generation;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeVariableName;

import java.util.Arrays;
import java.util.Collection;

import kaufland.com.coachbasebinderapi.PersistenceConfig;
import kaufland.com.coachbasebinderapi.TypeConversion;

public class TypeConversionMethodsGeneration {

    public static final String READ_METHOD_NAME = "read";

    public static final String WRITE_METHOD_NAME = "write";

    private static final String GET_TYPE_CONVERSION_METHOD = "getInstance().getConnector().getTypeConversions";

    public Collection<MethodSpec> generate() {

        return Arrays.asList(
                MethodSpec.methodBuilder(READ_METHOD_NAME).
                addParameter(TypeName.OBJECT, "value").
                addParameter(Class.class, "clazz").
                addTypeVariable(TypeVariableName.get("T")).
                returns(TypeVariableName.get("T")).
                addCode(CodeBlock.builder().
                        addStatement("$N conversion = $N." + GET_TYPE_CONVERSION_METHOD + "().get(clazz)", TypeConversion.class.getCanonicalName(), PersistenceConfig.class.getCanonicalName()).
                        beginControlFlow("if(conversion == null)").
                        addStatement("return (T) value").
                        endControlFlow().
                        addStatement("return (T) conversion.read(value)").
                        build()).build(),

                MethodSpec.methodBuilder(WRITE_METHOD_NAME).
                        addParameter(TypeName.OBJECT, "value").
                        addParameter(Class.class, "clazz").
                        addTypeVariable(TypeVariableName.get("T")).
                        returns(TypeVariableName.get("T")).
                        addCode(CodeBlock.builder().
                                addStatement("$N conversion = $N." + GET_TYPE_CONVERSION_METHOD + "().get(clazz)", TypeConversion.class.getCanonicalName(), PersistenceConfig.class.getCanonicalName()).
                                beginControlFlow("if(conversion == null)").
                                addStatement("return (T) value").
                                endControlFlow().
                                addStatement("return (T) conversion.write(value)").
                                build()).build()

        );

    }
}

File: couchbase-entity/src/main/java/com/kaufland/generation/TypeConversionMethodsGeneration.kt
package com.kaufland.generation

import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.TypeVariableName

import java.util.Arrays

import kaufland.com.coachbasebinderapi.PersistenceConfig

class TypeConversionMethodsGeneration {

    fun generate(): Collection<FunSpec> {

        return Arrays.asList(
                FunSpec.builder(READ_METHOD_NAME).
                        addParameter("value", TypeUtil.any().copy(nullable = true)).
                        addParameter( "clazz", TypeUtil.classStar()).addTypeVariable(TypeVariableName.invoke("T")).
                        returns(TypeVariableName.invoke("T?")).addCode(CodeBlock.builder().
                        addStatement("val conversion = %T.$GET_TYPE_CONVERSION_METHOD().get(clazz)", PersistenceConfig::class).
                        beginControlFlow("if(conversion == null)").
                        addStatement("return value as T").
                        endControlFlow().
                        addStatement("return conversion?.read(value) as T").build()).build(),

                FunSpec.builder(WRITE_METHOD_NAME).addParameter("value", TypeUtil.any().copy(nullable = true)).
                        addParameter( "clazz", TypeUtil.classStar()).addTypeVariable(TypeVariableName.invoke("T")).
                        returns(TypeVariableName.invoke("T?")).
                        addCode(CodeBlock.builder().
                                addStatement("val conversion = %T.$GET_TYPE_CONVERSION_METHOD().get(clazz)", PersistenceConfig::class).
                                beginControlFlow("if(conversion == null)").addStatement("return value as T").
                                endControlFlow().addStatement("return conversion.write(value) as T").build()).build()

        )

    }

    companion object {

        val READ_METHOD_NAME = "read"

        val WRITE_METHOD_NAME = "write"

        private val GET_TYPE_CONVERSION_METHOD = "getInstance().getConnector().getTypeConversions"
    }
}
